src/ui/src/composables/useFilesEncoder/useFilesEncoder.ts

```diff
@@ -0,0 +1,114 @@
+import { MaybeRef, readonly, ref, toValue } from "vue";
+import { encodeFileAsDataURL } from "./encodeFileAsDataURL";
+
+export type UiFile = {
+	id: string;
+	name: string;
+	size: number;
+	file: File;
+};
+
+let fallbackId = 0;
+
+function convertFileToUiFile(file: File): UiFile {
+	const id = window.crypto?.randomUUID() ?? `file-${fallbackId++}`;
+
+	return {
+		id,
+		file,
+		name: file.name,
+		size: file.size,
+	};
+}
+
+export type EncodedFile = {
+	name: string;
+	type: string;
+	data: string;
+};
+
+export type UseFilesEncoderParams = {
+	multiple: MaybeRef<boolean>;
+};
+
+export function useFilesEncoder({ multiple }: UseFilesEncoderParams) {
+	const uiFiles = ref<UiFile[]>([]);
+
+	function calcTotalSize(files: readonly Pick<File, "size">[]): number {
+		return files.reduce((sum, file) => sum + file.size, 0);
+	}
+
+	function addFiles(files: File[]) {
+		const convertedFiles = files.map((file) => convertFileToUiFile(file));
+
+		uiFiles.value = toValue(multiple)
+			? uiFiles.value.concat(convertedFiles)
+			: convertedFiles.slice(0, 1);
+	}
+
+	function removeFile(id: string) {
+		const index = uiFiles.value.findIndex((uiFile) => uiFile.id === id);
+
+		if (index !== -1) {
+			uiFiles.value.splice(index, 1);
+		}
```

Comment:
If you use `shallowRef`
```suggestion
		uiFiles.value = uiFiles.value.filter((uiFile) => uiFile.id !== id);
```